File: lab-2.3/c++/AvengersTeam.cpp
#include "AvengersTeam.h"
#include "Ironman.h"
#include "DoctorStrange.h"
#include "Hulk.h"
#include "Widow.h"
#include <algorithm>

AvengersTeam::AvengersTeam() {}

AvengersTeam::AvengersTeam(const AvengersTeam& other) {
    for (const auto* member : other.members) {
        members.push_back(member->clone());
    }
}

AvengersTeam::~AvengersTeam() {
    for (auto* member : members) {
        delete member;
    }
}

void AvengersTeam::addMember(const IDescribable* member) {
    members.push_back(member->clone());
}

void AvengersTeam::removeMember(const std::string& name) {
    members.erase(
        std::remove_if(members.begin(), members.end(),
            [&name](const IDescribable* member) { return member->getName() == name; }),
        members.end()
    );
}

void AvengersTeam::removeBySuitModel(const std::string& suitModel) {
    members.erase(
        std::remove_if(members.begin(), members.end(),
            [&suitModel](const IDescribable* member) {
                const Ironman* ironman = dynamic_cast<const Ironman*>(member);
                return ironman && ironman->getSuitModel() == suitModel;
            }),
        members.end()
    );
}

void AvengersTeam::clearTeam() {
    for (auto* member : members) {
        delete member;
    }
    members.clear();
}

const IDescribable* AvengersTeam::getMostPowerful() const {
    if (members.empty()) return nullptr;
    const IDescribable* maxPower = members[0];
    int maxLevel = members[0]->getPower();
    for (const auto* member : members) {
        if (member->getPower() > maxLevel) {
            maxLevel = member->getPower();
            maxPower = member;
        }
    }
    return maxPower;
}

void AvengersTeam::capitalizeNames() {
    for (auto* member : members) {
        std::string name = member->getName();
        if (!name.empty()) {
            name[0] = std::toupper(name[0]);
            // Note: Потребує сеттера для зміни імені
        }
    }
}

const std::vector<IDescribable*>& AvengersTeam::getMembers() const {
    return members;
}

File: lab-2.3/c#/AvengersTeam.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CharacterLib
{
    public class AvengersTeam
    {
        private List<IDescribable> members;

        public AvengersTeam()
        {
            members = new List<IDescribable>();
        }

        public AvengersTeam(AvengersTeam other)
        {
            members = new List<IDescribable>(other.members);
        }

        public void AddMember(IDescribable member)
        {
            if (member is Ironman ironman)
            {
                members.Add(new Ironman(ironman));
            }
            else if (member is DoctorStrange strange)
            {
                members.Add(new DoctorStrange(strange));
            }
            else if (member is Hulk hulk)
            {
                members.Add(new Hulk(hulk));
            }
            else if (member is Widow widow)
            {
                members.Add(new Widow(widow));
            }
        }

        public void RemoveMember(string name)
        {
            members.RemoveAll(m => m.GetName() == name);
        }

        public void ClearTeam()
        {
            members.Clear();
        }

        public IDescribable GetMostPowerful()
        {
            if (!members.Any()) return null;
            return members.OrderByDescending(m => m.GetPower()).First();
        }

        public void CapitalizeNames()
        {
            foreach (var member in members)
            {
                string name = member.GetName();
                if (!string.IsNullOrEmpty(name))
                {
                    // Потребує сеттера
                }
            }
        }

        public IReadOnlyList<IDescribable> GetMembers() => members.AsReadOnly();
    }
}